proc/futex.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * phoenix-rtos
+ *
+ * operating system kernel
+ *
+ * futex
+ *
+ * copyright 2025 phoenix systems
+ * author: kamil kowalczyk
+ *
+ * this file is part of phoenix-rtos.
+ *
+ * %license%
+ */
+
+#include <stdatomic.h>
+#include <stdbool.h>
+#include "hal/hal.h"
+#include "lib/lib.h"
+#include "include/errno.h"
+#include "process.h"
+#include "threads.h"
+#include "futex.h"
+
+static futex_sleepqueue_t *proc_allocFutexSleepQueue(process_t *process, addr_t address)
+{
+	u32 i;
+	futex_sleepqueue_t *futex_sleepqueue = NULL;
+
+	/* Not found. Look for an free slot */
+	for (i = 0; i < FUTEX_SLEEPQUEUES_SIZE; i++) {
+		if (process->futex_sleepqueues[i].address == 0) {
+			futex_sleepqueue = &process->futex_sleepqueues[i];
+			futex_sleepqueue->address = address; /* reserve */
+			break;
+		}
+	}
+
+	return futex_sleepqueue;
+}
+
+static futex_sleepqueue_t *proc_getFutexSleepQueue(process_t *process, addr_t address)
+{
+	u32 key, idx, i;
+	futex_sleepqueue_t *futex_sleepqueue = NULL;
+
+	key = address >> 3;
+	key ^= key >> FUTEX_SLEEPQUEUES_BITS;
+	idx = key & FUTEX_SLEEPQUEUES_MASK;
+
+	/* Find a taken slot with the same address using linear probing */
+	i = idx;
+	while (1) {
+		if (process->futex_sleepqueues[i].address == address) {
+			futex_sleepqueue = &process->futex_sleepqueues[i];
+			break;
+		}
+		i = (i + 1) % FUTEX_SLEEPQUEUES_SIZE;
+		if (i == idx) {
+			break;
+		}
+	}
+
+	return futex_sleepqueue;
+}
+
+static bool proc_futexUnwait(futex_sleepqueue_t *sq, futex_waitctx_t *wc)
+{
+	bool r;
+	spinlock_ctx_t sc;
+
+	hal_spinlockSet(&sq->spinlock, &sc);
+	r = atomic_load(&wc->thread) != NULL;
+	if (r) {
+		LIST_REMOVE(&sq->waitctxs, wc);
+	}
+	hal_spinlockClear(&sq->spinlock, &sc);
+	return r;
+}
+
+int proc_futexWait(_Atomic(u32) *address, u32 value, time_t timeout)
+{
+	spinlock_ctx_t ctx;
+	futex_sleepqueue_t *sleepqueue;
+	futex_waitctx_t wc;
+	thread_t *current;
+	int err = EOK;
+	time_t total_timeout = 0;
+
+	current = proc_current();
+
+	proc_lockSet(&current->process->lock);
+	sleepqueue = proc_getFutexSleepQueue(current->process, (addr_t)address);
+	if (sleepqueue == NULL) {
+		sleepqueue = proc_allocFutexSleepQueue(current->process, (addr_t)address);
```

Comment:
It seems sleep queues are never freed. What happens if a futex based mutex gets destroyed?